File: apps/qizhi/mobile/MentorAI-Android/app/src/main/kotlin/com/mentorai/app/chat/ChatViewModel.kt
package com.mentorai.app.chat

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mentorai.app.networking.AuthError
import com.mentorai.app.support.decodingJsonEscapes
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.transformWhile
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID

/**
 * One per chat detail screen. Mirrors iOS `ChatViewModel` — load history, send + stream replies,
 * manage attachments, decode streamed JSON escapes when the stream ends.
 */
class ChatViewModel(
    val initialSessionId: String?,
    private val chatApi: ChatAPI,
    private val sessionApi: SessionAPI,
    private val attachmentApi: AttachmentAPI,
    private val tokenProvider: () -> String?,
) : ViewModel() {

    private val _session = MutableStateFlow<ChatSession?>(null)
    val session: StateFlow<ChatSession?> = _session.asStateFlow()

    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages.asStateFlow()

    private val _draft = MutableStateFlow("")
    val draft: StateFlow<String> = _draft.asStateFlow()

    private val _statusText = MutableStateFlow<String?>(null)
    val statusText: StateFlow<String?> = _statusText.asStateFlow()

    private val _isStreaming = MutableStateFlow(false)
    val isStreaming: StateFlow<Boolean> = _isStreaming.asStateFlow()

    private val _streamError = MutableStateFlow<String?>(null)
    val streamError: StateFlow<String?> = _streamError.asStateFlow()

    private val _isLoadingSession = MutableStateFlow(false)
    val isLoadingSession: StateFlow<Boolean> = _isLoadingSession.asStateFlow()

    private val _sessionError = MutableStateFlow<String?>(null)
    val sessionError: StateFlow<String?> = _sessionError.asStateFlow()

    private val _attachments = MutableStateFlow<List<ChatAttachment>>(emptyList())
    val attachments: StateFlow<List<ChatAttachment>> = _attachments.asStateFlow()

    private var streamJob: Job? = null
    private val uploadJobs = mutableMapOf<UUID, Job>()

    val currentSessionId: String? get() = _session.value?.id ?: initialSessionId

    val titleForDisplay: String
        get() = _session.value?.title?.takeIf { it.isNotBlank() }
            ?: if (initialSessionId == null) "新对话" else "对话"

    val canSend: Boolean
        get() {
            val hasText = _draft.value.trim().isNotEmpty()
            val hasDoneAttachment = _attachments.value.any { it.remotePath != null }
            val anyUploading = _attachments.value.any { it.isUploading }
            return !_isStreaming.value && !anyUploading && (hasText || hasDoneAttachment)
        }

    val isUploadingAnyAttachment: Boolean
        get() = _attachments.value.any { it.isUploading }

    fun setDraft(text: String) { _draft.value = text }
    fun setStreamError(msg: String?) { _streamError.value = msg }

    /** Non-suspending convenience for buttons (Retry, etc.). */
    fun retryLoad() {
        viewModelScope.launch { load() }
    }

    suspend fun load() {
        val id = initialSessionId ?: return
        val token = tokenProvider() ?: run {
            _sessionError.value = "未登录"
            return
        }
        _isLoadingSession.value = true
        _sessionError.value = null
        try {
            val detail = sessionApi.detail(id, token)
            _session.value = detail
            _messages.value = detail.messages
        } catch (err: AuthError) {
            _sessionError.value = err.errorDescription
        } catch (t: Throwable) {
            _sessionError.value = t.localizedMessage ?: "加载失败"
        } finally {
            _isLoadingSession.value = false
        }
    }

    fun send() {
        if (_isStreaming.value || isUploadingAnyAttachment) return
        val text = _draft.value.trim()
        val doneAttachments = _attachments.value.filter { it.remotePath != null }
        val donePaths = doneAttachments.mapNotNull { it.remotePath }
        if (text.isEmpty() && donePaths.isEmpty()) return
        val token = tokenProvider() ?: run {
            _streamError.value = "未登录"
            return
        }

        _draft.value = ""
        _streamError.value = null
        _statusText.value = null

        val userBubble = composeUserBubble(text, doneAttachments)
        _messages.update { it + ChatMessage(role = ChatRole.User, content = userBubble) }
        _messages.update { it + ChatMessage(role = ChatRole.Assistant, content = "") }
        val assistantIndex = _messages.value.size - 1
        _isStreaming.value = true

        val request = ChatSendRequest(
            // Match iOS: omit session_id when nil OR empty (coerce a blank id to null).
            sessionId = currentSessionId?.takeIf { it.isNotEmpty() },
            query = composeQuery(text, doneAttachments),
            filePaths = donePaths.ifEmpty { null },
            extraParams = null,
        )
        _attachments.value = emptyList()

        streamJob = viewModelScope.launch {
            try {
                // Mirror iOS: handle each event, then STOP once an End or Error frame arrives
                // (transformWhile emits the terminal event, then halts the collect) instead of
                // relying solely on the server closing the stream.
                chatApi.send(request, token)
                    .transformWhile { event ->
                        emit(event)
                        event !is ChatEvent.End && event !is ChatEvent.Error
                    }
                    .collect { event -> handle(event, assistantIndex) }
            } catch (err: AuthError) {
                _streamError.value = err.errorDescription
            } catch (t: Throwable) {
                _streamError.value = t.localizedMessage ?: "流式响应失败"
            } finally {
                decodeAssistantTrailingEscapes(assistantIndex)
                _isStreaming.value = false
                _statusText.value = null
                streamJob = null
            }
        }
    }

    fun cancel() {
        streamJob?.cancel()
        streamJob = null
        _isStreaming.value = false
        _statusText.value = null
    }

    fun addAttachment(context: Context, uri: Uri, filename: String, size: Long) {
        if (!AttachmentLimits.isAllowed(filename)) {
            _streamError.value = "不支持的文件类型：$filename"
            return
        }
        val attachment = ChatAttachment(
            localUri = uri,
            filename = filename,
            sizeBytes = size,
            status = ChatAttachment.Status.Uploading,
        )
        _attachments.update { it + attachment }
        val id = attachment.id
        val token = tokenProvider()
        if (token.isNullOrBlank()) {
            updateAttachmentStatus(id, ChatAttachment.Status.Error("未登录"))
            return
        }
        uploadJobs[id] = viewModelScope.launch {
            try {
                val path = attachmentApi.upload(context, uri, filename, token)
                updateAttachmentStatus(id, ChatAttachment.Status.Done(path))
            } catch (err: AuthError) {
                updateAttachmentStatus(id, ChatAttachment.Status.Error(err.errorDescription ?: "上传失败"))
            } catch (t: Throwable) {
                updateAttachmentStatus(id, ChatAttachment.Status.Error(t.localizedMessage ?: "上传失败"))
            } finally {
                uploadJobs.remove(id)
            }
        }
    }

    fun removeAttachment(id: UUID) {
        uploadJobs.remove(id)?.cancel()
        _attachments.update { list -> list.filterNot { it.id == id } }
    }

    private fun updateAttachmentStatus(id: UUID, status: ChatAttachment.Status) {
        _attachments.update { list ->
            list.map { if (it.id == id) it.copy(status = status) else it }
        }
    }

    private fun handle(event: ChatEvent, assistantIndex: Int) {
        when (event) {
            is ChatEvent.Start -> {
                if (_session.value == null) {
                    _session.value = ChatSession(id = event.sessionId)
                }
            }
            is ChatEvent.Loading -> if (event.text.isNotEmpty()) _statusText.value = event.text
            is ChatEvent.Thinking -> if (event.text.isNotEmpty()) _statusText.value = event.text
            is ChatEvent.Message -> {
                _statusText.value = null
                appendAssistantChunk(assistantIndex, event.content)
            }
            is ChatEvent.Card -> Unit
            is ChatEvent.Error -> _streamError.value = event.message
            ChatEvent.End -> _statusText.value = null
        }
    }

    private fun appendAssistantChunk(index: Int, chunk: String) {
        _messages.update { list ->
            if (index !in list.indices) return@update list
            val target = list[index]
            list.toMutableList().also { it[index] = target.copy(content = target.content + chunk) }
        }
    }

    /**
     * Backend streams the raw final_answer slice from the model's JSON; literal "\n", "\"" etc.
     * survive to the client. Decode the assembled message once when the stream ends so the
     * persisted bubble shows real newlines / quotes / etc. — mirrors the iOS fix.
     */
    private fun decodeAssistantTrailingEscapes(index: Int) {
        _messages.update { list ->
            if (index !in list.indices) return@update list
            val target = list[index]
            if (target.role != ChatRole.Assistant) return@update list
            val decoded = target.content.decodingJsonEscapes()
            if (decoded == target.content) return@update list
            list.toMutableList().also { it[index] = target.copy(content = decoded) }
        }
    }

    /** Mirrors iOS `composeQuery` — append `[filename]` markers so the planner sees the files. */
    private fun composeQuery(text: String, attachments: List<ChatAttachment>): String {
        if (attachments.isEmpty()) return text
        val markers = attachments.joinToString(" ") { "[${it.filename}]" }
        return if (text.isEmpty()) markers else "$text\n\n$markers"
    }

    /** Mirrors iOS `composeUserBubble` — emoji + filenames stacked above the user text. */
    private fun composeUserBubble(text: String, attachments: List<ChatAttachment>): String {
        if (attachments.isEmpty()) return text
        val header = attachments.joinToString("\n") { "📎 ${it.filename}" }
        return if (text.isEmpty()) header else "$header\n\n$text"
    }
}

File: qizhi/mobile/MentorAI-iOS/MentorAI/Chat/ChatViewModel.swift
import Foundation
import SwiftUI
import os

@MainActor
final class ChatViewModel: ObservableObject {
    @Published private(set) var session: ChatSession?
    @Published private(set) var messages: [ChatMessage] = []
    @Published var draft: String = ""
    @Published private(set) var statusText: String?
    @Published private(set) var isStreaming: Bool = false
    @Published var streamError: String?
    @Published private(set) var isLoadingSession: Bool = false
    @Published var sessionError: String?
    @Published private(set) var attachments: [ChatAttachment] = []

    let initialSessionID: String?
    private let chatAPI: ChatAPI
    private let sessionAPI: SessionAPI
    private let attachmentAPI: AttachmentAPI
    private let tokenProvider: () -> String?
    private var streamTask: Task<Void, Never>?
    private var uploadTasks: [UUID: Task<Void, Never>] = [:]
    private let log = Logger(subsystem: "com.mentorai.app", category: "Chat")

    init(sessionID: String?,
         chatAPI: ChatAPI,
         sessionAPI: SessionAPI,
         attachmentAPI: AttachmentAPI,
         tokenProvider: @escaping () -> String?) {
        self.initialSessionID = sessionID
        self.chatAPI = chatAPI
        self.sessionAPI = sessionAPI
        self.attachmentAPI = attachmentAPI
        self.tokenProvider = tokenProvider
    }

    var currentSessionID: String? {
        session?.id ?? initialSessionID
    }

    var titleForDisplay: String {
        if let session, !session.title.isEmpty { return session.title }
        if initialSessionID == nil { return "新对话" }
        return "对话"
    }

    var canSend: Bool {
        let hasText = !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let hasDoneAttachment = attachments.contains { $0.remotePath != nil }
        let anyUploading = attachments.contains { $0.isUploading }
        return !isStreaming && !anyUploading && (hasText || hasDoneAttachment)
    }

    var isUploadingAnyAttachment: Bool {
        attachments.contains { $0.isUploading }
    }

    func load() async {
        guard let id = initialSessionID else { return }
        guard let token = tokenProvider() else {
            sessionError = "未登录"
            return
        }
        isLoadingSession = true
        sessionError = nil
        defer { isLoadingSession = false }
        do {
            let detail = try await sessionAPI.detail(id: id, token: token)
            session = detail
            messages = detail.messages
            log.info("Loaded session \(id, privacy: .public) with \(detail.messages.count) messages")
        } catch let err as AuthError {
            sessionError = err.errorDescription
        } catch {
            sessionError = error.localizedDescription
        }
    }

    func send() {
        let text = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        let doneAttachments = attachments.filter { $0.remotePath != nil }
        let donePaths = doneAttachments.compactMap { $0.remotePath }
        guard !isStreaming, !isUploadingAnyAttachment, !text.isEmpty || !donePaths.isEmpty else { return }
        guard let token = tokenProvider(), !token.isEmpty else {
            streamError = "未登录"
            return
        }

        draft = ""
        streamError = nil
        statusText = nil

        let displayText = composeUserBubble(text: text, attachments: doneAttachments)
        messages.append(ChatMessage(role: .user, content: displayText))
        messages.append(ChatMessage(role: .assistant, content: ""))
        let assistantIndex = messages.count - 1
        isStreaming = true

        let request = ChatSendRequest(
            sessionID: currentSessionID,
            query: composeQuery(text: text, attachments: doneAttachments),
            filePaths: donePaths.isEmpty ? nil : donePaths,
            extraParams: nil
        )
        log.info("chat/send query.len=\(text.count, privacy: .public) files=\(donePaths.count, privacy: .public) paths=\(donePaths.joined(separator: "|"), privacy: .public)")

        attachments.removeAll()

        streamTask = Task { @MainActor [weak self] in
            guard let self else { return }
            defer {
                // Streamed content carries literal JSON escapes (\n, \" …) from the
                // backend; decode the finished message so it persists correctly after
                // the streaming-tail render-time decode stops applying.
                if assistantIndex < self.messages.count {
                    self.messages[assistantIndex].content =
                        self.messages[assistantIndex].content.decodingJSONEscapes()
                }
                self.isStreaming = false
                self.statusText = nil
                self.streamTask = nil
            }
            do {
                for try await event in self.chatAPI.send(request, token: token) {
                    if Task.isCancelled { break }
                    self.handle(event: event, assistantIndex: assistantIndex)
                    if case .end = event { return }
                    if case .error = event { return }
                }
            } catch is CancellationError {
                self.log.info("Chat stream cancelled")
            } catch let err as AuthError {
                self.streamError = err.errorDescription
                self.log.error("Chat stream failed: \(err.errorDescription ?? "?", privacy: .public)")
            } catch {
                self.streamError = error.localizedDescription
                self.log.error("Chat stream failed: \(error.localizedDescription, privacy: .public)")
            }
        }
    }

    func cancel() {
        streamTask?.cancel()
        streamTask = nil
        isStreaming = false
        statusText = nil
    }

    // MARK: Attachments

    func addAttachment(localURL: URL) {
        let filename = localURL.lastPathComponent
        guard AttachmentLimits.isAllowed(filename: filename) else {
            streamError = "不支持的文件类型：\(filename)"
            return
        }
        let size = (try? FileManager.default.attributesOfItem(atPath: localURL.path)[.size] as? Int64) ?? 0
        var attachment = ChatAttachment(localURL: localURL, filename: filename, sizeBytes: size, status: .uploading)
        attachments.append(attachment)
        let attachmentID = attachment.id

        guard let token = tokenProvider(), !token.isEmpty else {
            attachment.status = .error("未登录")
            updateAttachment(id: attachmentID, status: .error("未登录"))
            return
        }

        let task = Task { @MainActor [weak self] in
            guard let self else { return }
            defer { self.uploadTasks[attachmentID] = nil }
            do {
                let path = try await self.attachmentAPI.upload(fileURL: localURL, token: token)
                self.updateAttachment(id: attachmentID, status: .done(path: path))
                self.log.info("Uploaded attachment \(filename, privacy: .public) → \(path, privacy: .public)")
            } catch is CancellationError {
                self.removeAttachment(id: attachmentID)
            } catch let err as AuthError {
                self.updateAttachment(id: attachmentID, status: .error(err.errorDescription ?? "上传失败"))
            } catch {
                self.updateAttachment(id: attachmentID, status: .error(error.localizedDescription))
            }
        }
        uploadTasks[attachmentID] = task
    }

    func removeAttachment(id: UUID) {
        uploadTasks[id]?.cancel()
        uploadTasks[id] = nil
        attachments.removeAll { $0.id == id }
    }

    private func updateAttachment(id: UUID, status: ChatAttachment.Status) {
        guard let idx = attachments.firstIndex(where: { $0.id == id }) else { return }
        attachments[idx].status = status
    }

    /// Mirrors the web client: append each attachment's name as `[name]` so the
    /// text-only planner can detect the attachment and route to the right tool.
    /// The server never sees `file_paths` during planning, only this query string.
    private func composeQuery(text: String, attachments: [ChatAttachment]) -> String {
        guard !attachments.isEmpty else { return text }
        let fileNames = attachments.map { "[\($0.filename)]" }.joined(separator: " ")
        return text.isEmpty ? fileNames : "\(text)\n\n\(fileNames)"
    }

    private func composeUserBubble(text: String, attachments: [ChatAttachment]) -> String {
        guard !attachments.isEmpty else { return text }
        let lines = attachments.map { "📎 \($0.filename)" }
        let header = lines.joined(separator: "\n")
        if text.isEmpty { return header }
        return header + "\n\n" + text
    }

    private func handle(event: ChatEvent, assistantIndex: Int) {
        switch event {
        case .start(let sid):
            if session == nil {
                session = ChatSession(id: sid, title: "", createTime: "", updateTime: "")
            }
        case .loading(let text), .thinking(let text):
            if !text.isEmpty { statusText = text }
        case .message(let chunk):
            statusText = nil
            guard assistantIndex < messages.count else { return }
            messages[assistantIndex].content += chunk
        case .card:
            break
        case .error(let msg):
            streamError = msg
        case .end:
            statusText = nil
        }
    }
}
